pkg/repository/v1_repository_test.go

```diff
@@ -18,15 +18,74 @@ import (
 	"encoding/hex"
 	"fmt"
 	"io"
+	"os"
+	"path"
 	"strings"
 	"testing"
 
 	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/google/uuid"
+	"github.com/pingcap/tiup/pkg/localdata"
 	"github.com/pingcap/tiup/pkg/repository/crypto"
 	"github.com/pingcap/tiup/pkg/repository/v1manifest"
+	"github.com/pingcap/tiup/pkg/utils"
 	"github.com/stretchr/testify/assert"
 )
 
+// Create a profile directory
+// Contained stuff:
+//   - index.json: with wrong signature
+//   - snapshot.json: correct
+//   - tidb.json: correct
+//   - timestamp: with expired timestamp
+func genPollutedProfileDir() (string, error) {
+	uid := uuid.New().String()
+	dir := path.Join("/tmp", uid)
+
+	wd, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+
+	if err := utils.Copy(path.Join(wd, "testdata", "polluted"), dir); err != nil {
+		return "", err
+	}
+
+	return dir, nil
+}
+
+func TestPollutedManifest(t *testing.T) {
+	profileDir, err := genPollutedProfileDir()
+	assert.Nil(t, err)
+	defer os.RemoveAll(profileDir)
+
+	profile := localdata.NewProfile(profileDir, &localdata.TiUPConfig{})
+	local, err := v1manifest.NewManifests(profile)
+	assert.Nil(t, err)
+
+	// Mock remote expired
+	mirror := MockMirror{
+		Resources: map[string]string{
+			"/timestamp.json": `{"signatures":[{"keyid":"66d4ea1da00076c822a6e1b4df5eb1e529eb38f6edcedff323e62f2bfe3eaddd","sig":"V8MgDDCmfVb8N0O3unbAno8q6i2Ag1Sbr/3n12Odk8McKzZaif7OcDm1IZB5J3o7ajsBF1tduTrcO7OijJQvx8l9i6aZi9J1lb/eJpYsyvQWdzd/T7osdRkEIhtM4/sGFjGslOolTFmpA/U5IkJ+FWAi38YaFPRn8bfIPLGniRAYs4/qjLBB3RgBUlDIIVvTiJIHEHtf3Bqb5LjpEjW4XhmDK94LJbKUqfO/6oDnQzI6Rot7zBWwDQVrIHakvQxoqA5c2jtMHCXSdX9cN7aRrNO4csggMzvQot7K0JYYszlroXnsL2ioNMgcPhtoEaMLW9mFjmdgR0j1//n1mxtdWA=="}],"signed":{"_type":"timestamp","expires":"2000-08-01T14:47:48+08:00","meta":{"/snapshot.json":{"hashes":{"sha256":"24c9fa83f15eda0683999b98ac0ff87fb95aed91c10410891fb38313f38e35c1"},"length":1760}},"spec_version":"0.1.0","version":639}}`,
+		},
+	}
+	repo := NewV1Repo(&mirror, Options{}, local)
+	err = repo.UpdateComponentManifests()
+	assert.NotNil(t, err)
+	assert.Contains(t, err.Error(), "has expired at")
+
+	// Mock remote correct but local throw sign error
+	mirror = MockMirror{
+		Resources: map[string]string{
+			"/timestamp.json": `{"signatures":[{"keyid":"66d4ea1da00076c822a6e1b4df5eb1e529eb38f6edcedff323e62f2bfe3eaddd","sig":"rFJ0oAi94FDtGd/nDVddDaZUqRc3XmG/b2EWndvGKcJ/0lkJPrZY5KeXCPcwwaDWbFHqhVPPOV3VlOU/CF65tW1TJaNrhQgplnVzwdJxD0qUIWhiad5xmdLZ2XDAJP5bgyUEawqlmXhAAXN9X5jOLwW5WJehkn83ArF+laZ16mnjixhicg1E1OQNb7fvvAXNnKPwjJ3sn+si82IllCe2gvVCmTnKvAef2ujMf3T13DeGS2tM5n2kr9yBFCwW28lTzKAZeUgIpizlm+ihwckQl7sJf6x8ea8eGi68WGGEBuvXjBvgz+Q8lrj5X2VL00OYWxmxgjYctwGrRP9aVw7WvA=="}],"signed":{"_type":"timestamp","expires":"2020-08-01T14:47:48+08:00","meta":{"/snapshot.json":{"hashes":{"sha256":"24c9fa83f15eda0683999b98ac0ff87fb95aed91c10410891fb38313f38e35c1"},"length":1760}},"spec_version":"0.1.0","version":639}}`,
```

Comment:
"expires":"2020-08-01T14:47:48+08:00"

Reply:
I have modified it in https://github.com/pingcap/tiup/pull/644